connectivity/mqtt/client_common.c

```diff
@@ -0,0 +1,519 @@
+  }
+
+  cfg->general_config->port = -1;
+  cfg->general_config->qos = 1;
+  cfg->general_config->max_inflight = 20;
+  cfg->general_config->keepalive = 60;
+  cfg->general_config->clean_session = true;
+  cfg->general_config->protocol_version = MQTT_PROTOCOL_V311;
+  cfg->general_config->client_type = client_type;
+}
+
+void mosq_config_free(mosq_config_t *cfg) {
+  if (cfg->general_config->client_type == client_pub || cfg->general_config->client_type == client_duplex) {
+    free(cfg->pub_config->message);
+    free(cfg->pub_config->topic);
+    free(cfg->pub_config->response_topic);
+  }
+  if ((cfg->general_config->client_type == client_sub) || (cfg->general_config->client_type == client_duplex)) {
+    if (cfg->sub_config->topics) {
+      for (int i = 0; i < cfg->sub_config->topic_count; i++) {
+        free(cfg->sub_config->topics[i]);
+      }
+      free(cfg->sub_config->topics);
+    }
+    if (cfg->sub_config->unsub_topics) {
+      for (int i = 0; i < cfg->sub_config->unsub_topic_count; i++) {
+        free(cfg->sub_config->unsub_topics[i]);
+      }
+      free(cfg->sub_config->unsub_topics);
+    }
+  }
+  free(cfg->general_config->id);
+  free(cfg->general_config->host);
+  // free(cfg->general_config->bind_address);
+  free(cfg->general_config->username);
+  free(cfg->general_config->password);
+  free(cfg->general_config->will_topic);
+  free(cfg->general_config->will_payload);
+
+#ifdef WITH_TLS
+  free(cfg->tls_config->cafile);
+  free(cfg->tls_config->capath);
+  free(cfg->tls_config->certfile);
+  free(cfg->tls_config->keyfile);
+  free(cfg->tls_config->ciphers);
+  free(cfg->tls_config->tls_alpn);
+  free(cfg->tls_config->tls_version);
+  free(cfg->tls_config->tls_engine);
+  free(cfg->tls_config->tls_engine_kpass_sha1);
+  free(cfg->tls_config->keyform);
+#ifdef FINAL_WITH_TLS_PSK
+  free(cfg->tls_config->psk);
+  free(cfg->tls_config->psk_identity);
+#endif
+#endif
+
+#ifdef WITH_SOCKS
+  free(cfg->socks_config->socks5_host);
+  free(cfg->socks_config->socks5_username);
+  free(cfg->socks_config->socks5_password);
+#endif
+  mosquitto_property_free_all(&cfg->property_config->connect_props);
+  mosquitto_property_free_all(&cfg->property_config->publish_props);
+  mosquitto_property_free_all(&cfg->property_config->subscribe_props);
+  mosquitto_property_free_all(&cfg->property_config->unsubscribe_props);
+  mosquitto_property_free_all(&cfg->property_config->disconnect_props);
+  mosquitto_property_free_all(&cfg->property_config->will_props);
+}
+
+status_t cfg_add_topic(mosq_config_t *cfg, client_type_t client_type, char *topic) {
+  if (cfg == NULL) {
+    return SC_MQTT_NULL;
+  }
+
+  if (mosquitto_validate_utf8(topic, strlen(topic))) {
+    fprintf(stderr, "Error: Malformed UTF-8 in topic argument.\n\n");
+    return SC_MQTT_TOPIC_SET;
+  }
+  if (client_type == client_pub || client_type == client_duplex) {
+    if (mosquitto_pub_topic_check(topic) == MOSQ_ERR_INVAL) {
```

Comment:
Move this check out of the `if...else` statement, since you check `topic` no matter which branch is into.
```
if (mosquitto_pub_topic_check(topic) == MOSQ_ERR_INVAL) {
...
}

if (client_type == client_pub || client_type == client_duplex) {
    cfg->pub_config->topic = strdup(topic);
} else if () {
....

```